amethyst_renderer/src/pipe/pass.rs

```diff
@@ -3,174 +3,60 @@
+    fn apply(&self, enc: &mut Encoder, effect: &mut Effect, scene: &Scene, model: &Model);
 }
 
-/// Simple prepparation pass
-#[derive(Clone)]
-pub struct BasicPass(BasicFn);
-
-/// Simple-processing pass
-#[derive(Clone)]
-pub struct SimplePass(SimpleFn, Effect);
-
-/// Model pass renders each model
 #[derive(Clone)]
-pub struct ModelPass(ModelFn, Effect);
-
-/// Model pass renders each light
-#[derive(Clone)]
-pub struct LightPass(LightFn, Effect);
-
-impl BasicPass {
-    /// Applies the rendering pass using the given `Encoder` and `Target`.
-    pub fn apply(&self, enc: &mut Encoder, out: &Target) {
-        (self.0)(enc, out)
-    }
-}
-
-impl Debug for BasicPass {
-    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("BasicPass")
-            .field(&"[closure]")
-            .finish()
-    }
-}
-
-impl SimplePass {
-    /// Applies the rendering pass using the given `Encoder`, `Target` and `Scene`.
-    pub fn apply(&self, enc: &mut Encoder, out: &Target, scene: &Scene) {
-        (self.0)(enc, out, &self.1, scene)
-    }
-}
-
-impl Debug for SimplePass {
-    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("SimplePass")
-            .field(&"[closure]")
-            .field(&self.1)
-            .finish()
-    }
-}
-
-impl ModelPass {
-    /// Applies the rendering pass using the given `Encoder`, `Target`, `Scene` and `Model`.
-    pub fn apply(&self, enc: &mut Encoder, out: &Target, scene: &Scene, model: &Model) {
-        (self.0)(enc, out, &self.1, scene, model)
-    }
-}
+pub(crate) struct Description(Arc<Pass>);
 
-impl Debug for ModelPass {
-    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("ModelPass")
-            .field(&"[closure]")
-            .field(&self.1)
-            .finish()
+impl Description {
+    pub fn new<P: Pass + 'static>(pass: P) -> Self {
+        Description(Arc::new(pass))
     }
-}
 
-impl LightPass {
-    /// Applies the rendering pass using the given `Encoder`, `Target`, `Scene` and `Light`.
-    pub fn apply(&self, enc: &mut Encoder, out: &Target, scene: &Scene, light: &Light) {
-        (self.0)(enc, out, &self.1, scene, light)
+    pub fn compile(self, fac: &mut Factory, out: &Target) -> Result<CompiledPass> {
+        let eb = NewEffect::new(fac, out);
+        let effect = self.0.compile(eb)?;
+        Ok(CompiledPass {
+            effect,
+            inner: self.0,
+        })
     }
 }
 
-impl Debug for LightPass {
+impl Debug for Description {
     fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("LightPass")
-            .field(&"[closure]")
-            .field(&self.1)
+        fmt.debug_tuple("PassDesc")
+            .field(&"[impl]")
             .finish()
     }
 }
 
 #[derive(Clone)]
-pub enum PassBuilder<'a> {
-    Basic(BasicFn),
-    Simple(SimpleFn, EffectBuilder<'a>),
-    Model(ModelFn, EffectBuilder<'a>),
-    Light(LightFn, EffectBuilder<'a>),
+pub struct CompiledPass {
```

Comment:
Deferred shading couldn't be implemented with only model passes.
I'm preparing PR to allow `Pass` to work on different kinds of components. It requires for new render to be integrated with ECS first. ECS types will be referenced only in `RenderSystem` ofc.